ionic-team/stencil
# fix(testing): puppeteer v10 support

## Important Front Matter

This PR is a port of https://github.com/ionic-team/stencil/pull/2934, which has largely already been reviewed:
- https://github.com/ionic-team/stencil/commit/63cc794387ec037519eb89cbb70fbbe80184b39d is a cherry-pick of the feature as landed in the v3.0.0-dev branch, https://github.com/ionic-team/stencil/commit/0c09aaacca8a1164e254edc58a7cc0357fb42fde
- https://github.com/ionic-team/stencil/commit/d01af05262faa2220e7c1b4788a1bea6fdd9f26c is the commit we [reverted in the v3.0.0-dev branch](https://github.com/ionic-team/stencil/pull/2934/commits/685f96c988bfc0864a9d75c2c9ec5a73758aa4d3)

The 'new stuff' includes:
- https://github.com/ionic-team/stencil/commit/d8ac36aa80d700b921a50a7a095bf2be8f1236cc - Lars' suggestions for merging declarations
- A minor package-lock fix

## Pull request checklist

Please check if your PR fulfills the following requirements:
   - Manual testing strategy described below

## Pull request type

 

 

Please check the type of change your PR introduces:
- [X] Other (please describe): Backport of https://github.com/ionic-team/stencil/pull/2934


## What is the current behavior?


In this PR, we upgrade Stencil to allow support for v10.0.0 of Puppeteer. The version of Puppeteer (v5.5.0) supported in `stencil@2.6.0` has two separate issues:
1. `npm install` cannot be easily run in the Stencil repo itself on newer Mac machines (those using an M1). This was fixed on the Puppeteer side in https://github.com/puppeteer/puppeteer/pull/7099
2. Existing clients cannot upgrade their version of Puppeteer


## What is the new behavior?


- Allows v10.0.0 of Puppeteer to be installed when developing for this repo and in downstream projects
- Removes the usage of `@types/puppeteer`
  - In v6 of Puppeteer, the library moved from using DefinitelyTyped community sourced type declarations to generating/exporting their own
  - As a result, we must update Stencil code to conform to these new typings
  - Despite this, I don't believe this causes a _breaking_ change, since our external API for testing is a facade in front of Puppeteer's
- Updates internal typings to conform with Puppeteer

## Does this introduce a breaking change?




## Manual Testing

- Ran `npm build` and `npm pack` from the root of the repo directory (on this branch)
- Spun up a boilerplate Stencil app using the [Stencil CLI](https://stenciljs.com/docs/getting-started) `npm init stencil`
- Installed the created tarball for my project 
- Installed the necessary test deps, and ran `npm t` in my stencil app passed 
- Dove 'into' the typings with both Puppeteer v5.5.0 and v10.0.0 to ensure there weren't any regressions with the type system/developer experience 
 
## Other information

- I plan on squashing this branch when it's time to merge. I'll update the commit body at that time
